### PR TITLE
use cache for referential integrity and escalation checks

### PR DESCRIPTION
**Pull request by deads2k**
Removes `ReadOnlyClusterPolicy` entirely.  The simulated registries are contained to the backing `RESTStorage` with no direct access.  The `RESTStorage` is scoped to the API installer path, so any other access comes via the API.

But the performance problem after that is:

```
        0     0% 0.033%    9202378 36.34%  github.com/openshift/origin/pkg/authorization/registry/clusterpolicy.(*storage).GetClusterPolicy
     14659 0.058% 0.091%    8639500 34.11%  k8s.io/kubernetes/pkg/runtime/serializer/json.(*Serializer).Decode
         0     0% 0.091%    8523402 33.66%  github.com/ugorji/go/codec.(*Decoder).Decode
         0     0% 0.091%    8523402 33.66%  github.com/ugorji/go/codec.(*Decoder).decode
         0     0% 0.091%    8348217 32.96%  github.com/ugorji/go/codec.(*Decoder).decodeI
   1599096  6.31%  6.41%    8331261 32.90%  github.com/ugorji/go/codec.(*decFnInfo).kStruct
         0     0%  6.41%    8199697 32.38%  github.com/ugorji/go/codec.(*Decoder).decodeValue
      5773 0.023%  6.43%    7907543 31.22%  k8s.io/kubernetes/pkg/runtime/serializer/recognizer.(*decoder).Decode
         0     0%  6.43%    7747725 30.59%  github.com/ugorji/go/codec.(*decFnInfo).kSlice
   5990754 23.66% 30.08%    6015374 23.75%  reflect.Value.call
         0     0% 30.08%    6005989 23.72%  reflect.Value.Call
         0     0% 30.08%    5822174 22.99%  github.com/openshift/origin/pkg/authorization/registry/clusterpolicy.ReadOnlyClusterPolicy.Get
     32485  0.13% 30.21%    5470939 21.60%  k8s.io/kubernetes/pkg/conversion.(*Converter).doConversion
     21630 0.085% 30.30%    5437367 21.47%  k8s.io/kubernetes/pkg/conversion.(*Converter).Convert
     16419 0.065% 30.36%    5327760 21.04%  k8s.io/kubernetes/pkg/conversion.(*Converter).defaultConvert
      2048 0.0081% 30.37%    5300805 20.93%  k8s.io/kubernetes/pkg/conversion.(*Converter).convertKV
         0     0% 30.37%    5214307 20.59%  k8s.io/kubernetes/pkg/conversion.(*Converter).DefaultConvert
         0     0% 30.37%    5214307 20.59%  k8s.io/kubernetes/pkg/conversion.(*scope).DefaultConvert
```

The dominator call to GetClusterPolicy

```
                                           5818843 63.40% |   github.com/openshift/origin/pkg/authorization/registry/clusterpolicy.ReadOnlyClusterPolicy.Get
                                           3359638 36.60% |   github.com/openshift/origin/pkg/authorization/registry/clusterpolicy.(*simulatedStorage).GetPolicy
         0     0% 0.0051%    9202378 36.34%                | github.com/openshift/origin/pkg/authorization/registry/clusterpolicy.(*storage).GetClusterPolicy
                                           9202378   100% |   k8s.io/kubernetes/pkg/registry/generic/registry.(*Store).Get
```

### CODE REVIEW

**Comment by deads2k** [comment]
[test]

**Comment by deads2k** [comment]
@smarterclayton this would be the one to try next.  If its still a problem, I'll start inspecting stack dumps and panic-ing to chase it further.

File: pkg/controller/authorization_informers.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
+package controller
```
Comment (smarterclayton) [inline pkg/controller/authorization_informers.go]: Would really prefer this to be in its own package.

Comment (deads2k) [inline pkg/controller/authorization_informers.go]: > Would really prefer this to be in its own package.

@kargakis agreed to move it in a followup.

Comment (smarterclayton) [inline pkg/controller/authorization_informers.go]: Fine with me.  Still looking.

File: pkg/controller/authorization_informers.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+	"k8s.io/kubernetes/pkg/watch"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	"github.com/openshift/origin/pkg/client"
+	ocache "github.com/openshift/origin/pkg/client/cache"
+)
+
+type ClusterPolicyInformer interface {
+	Informer() framework.SharedIndexInformer
+	// still use an indexer, no telling what someone will want to index on someday
```
Comment (smarterclayton) [inline pkg/controller/authorization_informers.go]: Your godoc'ing is slipping...  fine for now.

File: pkg/controller/authorization_informers.go [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+func (f *policyInformer) Indexer() cache.Indexer {
+	informer := f.Informer()
+	return informer.GetIndexer()
+}
+
+func (f *policyInformer) Lister() client.SyncedPoliciesListerNamespacer {
+	return &ocache.InformerToPolicyNamespacer{SharedIndexInformer: f.Informer()}
+}
+
+type PolicyBindingInformer interface {
```
Comment (smarterclayton) [inline pkg/controller/authorization_informers.go]: This interface construct could be better - it's not a great interface, and it doesn't really hide the differences much.

**Comment by liggitt** [comment]
last commit LGTM

**Comment by smarterclayton** [comment]
LGTM beyond the other comments.  Merge away.

**Comment by deads2k** [comment]
[merge]

**Pull request merged**